File: mobile/test/features/notification/domain/review_deep_link_test.dart
// Unit tests for the mobile review deep-link helper.
//
// Mirrors the web vitest coverage for buildReviewDeepLink — every
// review notification kind produces a /chat/<conv>?openReview=1&
// reviewProposalId=<id> path, every other kind returns null, and
// malformed payloads (missing/empty/non-string fields, legacy
// dispute_id-only payloads) silently degrade to null instead of
// throwing.

import 'package:flutter_test/flutter_test.dart';
import 'package:marketplace_mobile/features/notification/domain/entities/app_notification.dart';
import 'package:marketplace_mobile/features/notification/domain/review_deep_link.dart';

AppNotification _make(String type, Map<String, dynamic> data) {
  return AppNotification(
    id: 'n-1',
    userId: 'u-1',
    type: type,
    title: 'title',
    body: 'body',
    data: data,
    createdAt: DateTime.utc(2026, 5, 12),
  );
}

void main() {
  group('buildReviewDeepLink', () {
    const proposalId = '11111111-1111-1111-1111-111111111111';
    const conversationId = '22222222-2222-2222-2222-222222222222';

    test('proposal_completed with full payload → returns the deep link', () {
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'proposal_id': proposalId,
        'conversation_id': conversationId,
        'proposal_title': 'Site web Acme',
      }));

      expect(link, isNotNull);
      expect(link!.conversationId, conversationId);
      expect(link.proposalId, proposalId);
      expect(
        link.toRoute('/chat'),
        '/chat/$conversationId?openReview=1&reviewProposalId=$proposalId',
      );
    });

    test('review_received with full payload → returns the deep link', () {
      final link = buildReviewDeepLink(_make('review_received', {
        'proposal_id': proposalId,
        'conversation_id': conversationId,
      }));

      expect(link, isNotNull);
      expect(link!.proposalId, proposalId);
    });

    test('non-review kinds → returns null', () {
      const nonReviewKinds = <String>[
        'proposal_received',
        'proposal_accepted',
        'proposal_declined',
        'proposal_modified',
        'proposal_paid',
        'completion_requested',
        'new_message',
        'system_announcement',
      ];
      for (final kind in nonReviewKinds) {
        final link = buildReviewDeepLink(_make(kind, {
          'proposal_id': proposalId,
          'conversation_id': conversationId,
        }));
        expect(link, isNull, reason: 'kind $kind must not deep-link');
      }
    });

    test('missing proposal_id → null', () {
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'conversation_id': conversationId,
      }));
      expect(link, isNull);
    });

    test('missing conversation_id → null', () {
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'proposal_id': proposalId,
      }));
      expect(link, isNull);
    });

    test('legacy dispute payload (only dispute_id) → null (regression guard)', () {
      // Pre-fix, dispute resolution notifications shipped only with
      // dispute_id. A stale notification arriving post-fix must not
      // crash the tap handler — it must silently fall back to
      // "mark-as-read only".
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'dispute_id': 'd-1',
      }));
      expect(link, isNull);
    });

    test('empty string values → null', () {
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'proposal_id': '',
        'conversation_id': conversationId,
      }));
      expect(link, isNull);
    });

    test('non-string values → null', () {
      final link = buildReviewDeepLink(_make('proposal_completed', {
        'proposal_id': 42,
        'conversation_id': conversationId,
      }));
      expect(link, isNull);
    });
  });
}

File: web/src/features/notification/lib/__tests__/review-deep-link.test.ts
import { describe, it, expect } from "vitest"
import { buildReviewDeepLink } from "../review-deep-link"
import type { Notification, NotificationType } from "../../types"

function makeNotification(
  type: NotificationType,
  data: Record<string, unknown>,
): Notification {
  return {
    id: "notif-1",
    user_id: "user-1",
    type,
    title: "Whatever",
    body: "Whatever",
    data,
    read_at: null,
    created_at: new Date().toISOString(),
  }
}

const PROPOSAL_ID = "11111111-1111-1111-1111-111111111111"
const CONVERSATION_ID = "22222222-2222-2222-2222-222222222222"

describe("buildReviewDeepLink", () => {
  describe("review-related notification kinds", () => {
    const reviewKinds: NotificationType[] = [
      "proposal_completed",
      "review_received",
    ]

    it.each(reviewKinds)(
      "builds /messages?id=…&openReview=1&reviewProposalId=… for kind %s",
      (kind) => {
        const url = buildReviewDeepLink(
          makeNotification(kind, {
            proposal_id: PROPOSAL_ID,
            conversation_id: CONVERSATION_ID,
            proposal_title: "Site web Acme",
          }),
        )
        expect(url).not.toBeNull()
        const parsed = new URL(`http://x${url!}`)
        expect(parsed.pathname).toBe("/messages")
        expect(parsed.searchParams.get("id")).toBe(CONVERSATION_ID)
        expect(parsed.searchParams.get("openReview")).toBe("1")
        expect(parsed.searchParams.get("reviewProposalId")).toBe(PROPOSAL_ID)
      },
    )
  })

  describe("non-review notification kinds", () => {
    const otherKinds: NotificationType[] = [
      "proposal_received",
      "proposal_accepted",
      "proposal_declined",
      "proposal_modified",
      "proposal_paid",
      "completion_requested",
      "new_message",
      "system_announcement",
    ]

    it.each(otherKinds)(
      "returns null for kind %s (no deep-link)",
      (kind) => {
        const url = buildReviewDeepLink(
          makeNotification(kind, {
            proposal_id: PROPOSAL_ID,
            conversation_id: CONVERSATION_ID,
          }),
        )
        expect(url).toBeNull()
      },
    )
  })

  it("returns null when proposal_id is missing", () => {
    const url = buildReviewDeepLink(
      makeNotification("proposal_completed", {
        conversation_id: CONVERSATION_ID,
      }),
    )
    expect(url).toBeNull()
  })

  it("returns null when conversation_id is missing", () => {
    const url = buildReviewDeepLink(
      makeNotification("proposal_completed", {
        proposal_id: PROPOSAL_ID,
      }),
    )
    expect(url).toBeNull()
  })

  it("returns null when data is empty (legacy dispute payload)", () => {
    // Regression guard: pre-fix, dispute resolution notifications
    // shipped only { dispute_id: "..." } — the new payload aligns
    // with the proposal flow, but in case an older client still
    // surfaces a stale notification we should not crash, just no-op.
    const url = buildReviewDeepLink(
      makeNotification("proposal_completed", {
        dispute_id: "00000000-0000-0000-0000-000000000000",
      }),
    )
    expect(url).toBeNull()
  })

  it("returns null when data is malformed", () => {
    const url = buildReviewDeepLink(
      makeNotification("proposal_completed", {
        proposal_id: 12345,
        conversation_id: { x: 1 },
      }),
    )
    expect(url).toBeNull()
  })
})
